--- /dev/null
+++ ./../src/main/java/com/lushprojects/circuitjs1/client/circuitjs1.java
@@ -76,7 +76,7 @@
 	public static final String versionString="2.7.1js";
 	
 	// Set to true if the server runs the shortrelay.php file in the same directory as the circuit simulator
-	public static final boolean shortRelaySupported = true;
+	public static final boolean shortRelaySupported = false;
 
 	static CirSim mysim;
 	HashMap<String,String> localizationMap;
