--- /dev/null
+++ src/main/java/com/lushprojects/circuitjs1/client/circuitjs1.java
@@ -40,7 +40,7 @@
 
     // Set to true if the server runs the shortrelay.php file in the same directory
     // as the circuit simulator
-    public static final boolean shortRelaySupported = true;
+    public static final boolean shortRelaySupported = false;
 
     static CirSim mysim;
 
@@ -153,7 +153,7 @@
         Window.addResizeHandler(new ResizeHandler() {
             public void onResize(ResizeEvent event) {
                 mysim.setCanvasSize();
-                mysim.setiFrameHeight();
+                mysim.setSlidersPanelHeight();
             }
         });
 
