--- /dev/null
+++ src/main/java/com/lushprojects/circuitjs1/client/ExportAsUrlDialog.java
@@ -45,8 +45,6 @@
 	String requrl;
 	
 	public boolean shortIsSupported() {
-		if (CirSim.theSim.isElectron())
-		    return false;
 		return circuitjs1.shortRelaySupported;
 	}
 	
@@ -95,11 +93,9 @@
 	public ExportAsUrlDialog( String dump) {
 		super();
 		closeOnEnter = false;
-		String start[] = Location.getHref().split("\\?");
-		if (CirSim.theSim.isElectron())
-		    start[0] = "https://www.falstad.com/circuit/circuitjs.html";
+		String start = "https://www.falstad.com/circuit/circuitjs.html";
 		String query="?ctz=" + compress(dump);
-		dump = start[0] + query;
+		dump = start + query;
 		requrl = URL.encodeQueryString(query);
 		Button okButton, copyButton;
 	
