--- /dev/null
+++ src/main/java/com/lushprojects/circuitjs1/client/Scrollbar.java
@@ -78,9 +78,9 @@
 		val=value;
 		 pan = new VerticalPanel();
 		can = Canvas.createIfSupported();
-		can.setWidth((CirSim.VERTICALPANELWIDTH)+" px");
+		can.setWidth("150 px");
 		can.setHeight("40 px");
-		can.setCoordinateSpaceWidth(CirSim.VERTICALPANELWIDTH);
+		can.setCoordinateSpaceWidth(150);
 		can.setCoordinateSpaceHeight(SCROLLHEIGHT);
 		pan.add(can);
 		g=can.getContext2d();
@@ -121,23 +121,23 @@
 		else
 			g.setStrokeStyle("lightgrey");
 		g.setLineWidth(1.0);
-		g.fillRect(0,0,CirSim.VERTICALPANELWIDTH,SCROLLHEIGHT);
+		g.fillRect(0,0,150,SCROLLHEIGHT);
 		g.beginPath();
 		g.moveTo(HMARGIN+SCROLLHEIGHT-3, 0);
 		g.lineTo(HMARGIN, SCROLLHEIGHT/2);
 		g.lineTo(HMARGIN+SCROLLHEIGHT-3, SCROLLHEIGHT);
-		g.moveTo(CirSim.VERTICALPANELWIDTH-HMARGIN-SCROLLHEIGHT+3, 0);
-		g.lineTo(CirSim.VERTICALPANELWIDTH-HMARGIN, SCROLLHEIGHT/2);
-		g.lineTo(CirSim.VERTICALPANELWIDTH-HMARGIN-SCROLLHEIGHT+3, SCROLLHEIGHT);
+		g.moveTo(150-HMARGIN-SCROLLHEIGHT+3, 0);
+		g.lineTo(150-HMARGIN, SCROLLHEIGHT/2);
+		g.lineTo(150-HMARGIN-SCROLLHEIGHT+3, SCROLLHEIGHT);
 		g.stroke();
 		if (enabled)
 			g.setStrokeStyle("grey");
 		g.beginPath();
 		g.setLineWidth(5.0);
 		g.moveTo(HMARGIN+SCROLLHEIGHT+BARMARGIN, SCROLLHEIGHT/2);
-		g.lineTo(CirSim.VERTICALPANELWIDTH-HMARGIN-SCROLLHEIGHT-BARMARGIN, SCROLLHEIGHT/2);
+		g.lineTo(150-HMARGIN-SCROLLHEIGHT-BARMARGIN, SCROLLHEIGHT/2);
 		g.stroke();
-		double p=HMARGIN+SCROLLHEIGHT+BARMARGIN+((CirSim.VERTICALPANELWIDTH-2*(HMARGIN+SCROLLHEIGHT+BARMARGIN))*((double)(val-min)))/(max-min);
+		double p=HMARGIN+SCROLLHEIGHT+BARMARGIN+((150-2*(HMARGIN+SCROLLHEIGHT+BARMARGIN))*((double)(val-min)))/(max-min);
 		if (enabled) {
 			if (attachedElm!=null && attachedElm.needsHighlight())
 				g.setStrokeStyle(CircuitElm.selectColor.getHexValue());
@@ -163,7 +163,7 @@
 	
 	int calcValueFromPos(int x){
 		int v;
-		v= min+(max-min)*(x-HMARGIN-SCROLLHEIGHT-BARMARGIN)/(CirSim.VERTICALPANELWIDTH-2*(HMARGIN+SCROLLHEIGHT+BARMARGIN));
+		v= min+(max-min)*(x-HMARGIN-SCROLLHEIGHT-BARMARGIN)/(150-2*(HMARGIN+SCROLLHEIGHT+BARMARGIN));
 		if (v<min)
 			v=min;
 		if (v>max)
@@ -185,7 +185,7 @@
 			val--;
 		}
 		else {
-		    if (x > CirSim.VERTICALPANELWIDTH-HMARGIN-SCROLLHEIGHT ) {
+		    if (x > 150-HMARGIN-SCROLLHEIGHT ) {
 			if (val<max)
 			    val++;
 		    }
