--- /dev/null
+++ src/main/java/com/lushprojects/circuitjs1/client/Toolbar.java
@@ -13,6 +13,7 @@
 import com.google.gwt.user.client.Window;
 import com.google.gwt.user.client.ui.*;
 import com.lushprojects.circuitjs1.client.util.Locale;
+import com.google.gwt.user.client.ui.HTML;
 
 import java.util.HashMap;
 
@@ -21,6 +22,7 @@
     private Label modeLabel;
     private HashMap<String, Label> highlightableButtons = new HashMap<>();
     private Label activeButton;  // Currently active button
+    private String SEPARATOR = "<div style=\"height:30px;width:0;border-left:2px solid grey;\"></div>";
 
     Label resistorButton;
 
@@ -35,17 +37,28 @@
         style.setDisplay(Style.Display.FLEX);
 	setVerticalAlignment(ALIGN_MIDDLE);
 
+	add(createIconButton("doc-new", "New Blank Circuit", new MyCommand("file", "newblankcircuit")));
+	add(createIconButton("folder",  "Open File...", new MyCommand("file", "importfromlocalfile")));
+	add(createIconButton("floppy", "Save", new MyCommand("file", "save")));
+	add(createIconButton("floppy",  "Save As...", new MyCommand("file", "saveas")));
+	add(new HTML("<sup style=\"margin-left:-12px;\"><strong>+</strong></sup>")); // add symbol to "Save As" icon
+    add(new HTML(SEPARATOR));
 	add(createIconButton("ccw", "Undo", new MyCommand("edit", "undo")));
 	add(createIconButton("cw",  "Redo", new MyCommand("edit", "redo")));
+    add(new HTML(SEPARATOR));
 	add(createIconButton("scissors", "Cut", new MyCommand("edit", "cut")));
 	add(createIconButton("copy", "Copy", new MyCommand("edit", "copy")));
 	add(createIconButton("paste", "Paste", new MyCommand("edit", "paste")));
 	add(createIconButton("clone", "Duplicate", new MyCommand("edit", "duplicate")));
+    add(new HTML(SEPARATOR));
 	add(createIconButton("search", "Find Component...", new MyCommand("edit", "search")));
+    add(new HTML(SEPARATOR));
 
+	add(createIconButton("target", "Centre Circuit", new MyCommand("edit", "centrecircuit")));
 	add(createIconButton("zoom-11", "Zoom 100%", new MyCommand("zoom", "zoom100")));
 	add(createIconButton("zoom-in", "Zoom In", new MyCommand("zoom", "zoomin")));
 	add(createIconButton("zoom-out", "Zoom Out", new MyCommand("zoom", "zoomout")));
+    add(new HTML(SEPARATOR));
 
 	add(createIconButton(wireIcon, "WireElm"));
 	add(resistorButton = createIconButton(resistorIcon, "ResistorElm"));
@@ -74,7 +87,7 @@
 	String gateInfo[] = { andIcon, "AndGateElm", nandIcon, "NandGateElm", 
 			      orIcon, "OrGateElm", norIcon, "NorGateElm", xorIcon, "XorGateElm" };
 	add(createButtonSet(gateInfo));
-
+/*
         // Spacer to push the mode label to the right
         HorizontalPanel spacer = new HorizontalPanel();
         //spacer.style.setFlexGrow(1); // Fill remaining space
@@ -84,9 +97,10 @@
         modeLabel = new Label("");
         styleModeLabel(modeLabel);
         add(modeLabel);
+*/
     }
 
-    public void setModeLabel(String text) { modeLabel.setText(Locale.LS("Mode: ") + text); }
+    //public void setModeLabel(String text) { modeLabel.setText(Locale.LS("Mode: ") + text); }
 
     private Label createIconButton(String icon, String cls) {
 	CirSim sim = CirSim.theSim;
@@ -140,7 +154,7 @@
 
     String makeSvg(String s, int size) {
 	double scale = size/24.0;
-	return "<svg xmlns='http://www.w3.org/2000/svg' width='" + size + "' height='" + size + "'><g transform='scale(" + scale + ")'>" +
+	return "<svg width='" + size + "' height='" + size + "'><g transform='scale(" + scale + ")'>" +
                  s.substring(5, s.length()-5) + "<g></svg>";
     }
 
